infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/RestoreRedundancyMergeSegments.java

```diff
@@ -1,38 +1,224 @@
+            StateTransferManager transferManager, Layout currentLayout, long trimMark) {
+        // Create a pre transfer state list.
+        ImmutableList<TransferSegment> preTransferList =
+                redundancyCalculator.createStateList(currentLayout, trimMark);
+
+        // Perform a state transfer for each segment synchronously and update the state list.
+        ImmutableList<TransferSegment> transferList = transferManager
+                .handleTransfer(preTransferList);
+
+        // Get all the transfers that failed.
+        List<TransferSegment> failedList = transferList.stream()
+                .filter(segment -> segment.getStatus().getSegmentState() == FAILED)
+                .collect(Collectors.toList());
+
+        // Throw the first transfer segment exception if any of the transfers have failed.
+        Optional<TransferSegmentException> transferSegmentFailure = failedList.stream()
+                .findFirst()
+                .flatMap(ts -> ts.getStatus().getCauseOfFailure());
+
+        transferSegmentFailure.ifPresent(failure -> {
+            throw failure;
+        });
+
+        // Filter all the segments with a status TRANSFERRED.
+        return transferList.stream()
+                .filter(segment -> segment.getStatus().getSegmentState() == TRANSFERRED)
+                .collect(Collectors.toList());
+    }
+
     /**
-     * Returns set of nodes which are present in the next index but not in the specified segment. These
-     * nodes have reduced redundancy and state needs to be transferred only to these before these segments can be
-     * merged.
+     * Sets the trim mark on this endpoint's log unit, performs a prefix trim and then compaction.
      *
-     * @param layout             Current layout.
-     * @param layoutSegmentIndex Segment to compare to get nodes with reduced redundancy.
-     * @return Set of nodes with reduced redundancy.
+     * @param runtime A current runtime.
+     * @return A retrieved trim mark.
      */
-    private Set<String> getNodesWithReducedRedundancy(Layout layout, int layoutSegmentIndex) {
-        // Get the set of servers present in the next segment but not in the this
-        // segment.
-        return Sets.difference(
-                layout.getSegments().get(layoutSegmentIndex + 1).getAllLogServers(),
-                layout.getSegments().get(layoutSegmentIndex).getAllLogServers());
+    @VisibleForTesting
+    long trimLog(CorfuRuntime runtime) {
+
+        long trimMark = runtime.getAddressSpaceView().getTrimMark().getSequence();
+
+        Layout layout = runtime.getLayoutView().getLayout();
+
+        Token prefixToken = new Token(layout.getEpoch(), trimMark - 1);
+
+        LogUnitClient logUnitClient = runtime
+                .getLayoutView()
+                .getRuntimeLayout(layout)
+                .getLogUnitClient(currentNode);
+
+        logUnitClient.prefixTrim(prefixToken).join();
+
+        logUnitClient.compact().join();
```

Comment:
I added it as per Maithem's request

Reply:
I can't find my comment. But one reason where this would make sense, is if you transmit a large chunk of the log and fail at the end. Before retrying, you would want to "delete" (i.e. compact) all the addresses that have been already trimmed. Its an optimization that can help in situation where a node is getting rebuilt under low available space conditions.